### PR TITLE
Add `SurfaceTextureSurfaceProducer`

### PR DESCRIPTION
**Pull request by matanlurey**
Unblocks https://github.com/flutter/flutter/issues/139702.

The next step will be local validation that this API works for at least a few of our internal plugins (video_player, camera).

### CODE REVIEW

File: shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java [hunk 1]
```diff
@@ -159,21 +159,45 @@ private void clearDeadListeners() {
...
+    // Prior to Impeller, Flutter on Android *only* ran on OpenGLES (via Skia). That meant that
+    // plugins (i.e. end-users) either explicitly created a SurfaceTexture (via createX/registerX)
+    // or an ImageTexture (via createX/registerX).
+    //
+    // In an Impeller world, which for the first time uses (if available) a Vulkan rendering
+    // backend, it is no longer possible (at least not trivially) to render an OpenGLES-provided
+    // texture (SurfaceTexture) in a Vulkan context.
+    //
+    // This function picks the "best" rendering surface based on the Android runtime, and provides
+    // a backend-agnostic SurfaceProducer (which in turn vends a Surface), and has plugins (i.e.
```
Comment (johnmccutchan) [inline shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java:174]: consumer-agnostic

File: shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java [hunk 2]
```diff
@@ -159,21 +159,45 @@ private void clearDeadListeners() {
...
+    // end-users) use the Surface instead, letting us "hide" what the implementation is provided by.
```
Comment (johnmccutchan) [inline shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java:175]: letting us "hide" the consumer side of the implementation.

File: shell/platform/android/io/flutter/embedding/engine/renderer/SurfaceTextureSurfaceProducer.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import android.media.Image;
+import android.view.Surface;
+import androidx.annotation.NonNull;
+import io.flutter.view.TextureRegistry;
+
+// An incomplete implementation of TextureRegistry.SurfaceProducer.
+//
+// TODO(https://github.com/flutter/flutter/issues/139702).
+final class SurfaceTextureSurfaceProducer
+    implements TextureRegistry.SurfaceProducer, TextureRegistry.ImageConsumer {
```
Comment (johnmccutchan) [inline shell/platform/android/io/flutter/embedding/engine/renderer/SurfaceTextureSurfaceProducer.java:13]: This won't be an ImageConsumer, you'll have to introduce a new Consumer interface, maybe, "TextureRegistry.GLTextureConsumer".

File: shell/platform/android/io/flutter/embedding/engine/renderer/SurfaceTextureSurfaceProducer.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  public int getHeight() {
+    return 0;
+  }
+
+  @Override
+  public Surface getSurface() {
+    return new Surface(getSurfaceTexture());
+  }
+
+  @Override
```
Comment (johnmccutchan) [inline shell/platform/android/io/flutter/embedding/engine/renderer/SurfaceTextureSurfaceProducer.java:58]: Remove acquireLatestImage

File: shell/platform/android/io/flutter/embedding/engine/renderer/SurfaceTextureSurfaceProducer.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  public int getHeight() {
+    return requestedBufferHeight;
+  }
+
+  @Override
+  public Surface getSurface() {
+    return new Surface(getSurfaceTexture());
```
Comment (johnmccutchan) [inline shell/platform/android/io/flutter/embedding/engine/renderer/SurfaceTextureSurfaceProducer.java:57]: You should create the Surface once and cache it.

File: shell/platform/android/io/flutter/embedding/engine/renderer/SurfaceTextureSurfaceProducer.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
+package io.flutter.embedding.engine.renderer;
+
+import android.graphics.SurfaceTexture;
+import android.view.Surface;
+import androidx.annotation.NonNull;
+import io.flutter.view.TextureRegistry;
+
+/** Uses a {@link android.graphics.SurfaceTexture} to populate the texture registry. */
+final class SurfaceTextureSurfaceProducer
```
Comment (johnmccutchan) [inline shell/platform/android/io/flutter/embedding/engine/renderer/SurfaceTextureSurfaceProducer.java:9]: You need to add a finalizer (see the other impls)

File: shell/platform/android/io/flutter/embedding/engine/renderer/SurfaceTextureSurfaceProducer.java (hunk 5 above)
Comment (matanlurey) [reply shell/platform/android/io/flutter/embedding/engine/renderer/SurfaceTextureSurfaceProducer.java:57]: Ack! I was confused by this comment in the parent interface:

```java
/**
 * Get a Surface that can be used to update the texture contents.
 *
 * <p>NOTE: You should not cache the returned surface but instead invoke getSurface each time
 * you need to draw. The surface may change when the texture is resized or has its format
 * changed.
 *
 * @return a Surface to use for a drawing target for various APIs.
 */
Surface getSurface();
```

Perhaps we could modify to specify as a _consumer_ of this API (versus implementor)?

Comment (johnmccutchan) [reply shell/platform/android/io/flutter/embedding/engine/renderer/SurfaceTextureSurfaceProducer.java:57]: For SurfaceTexture you should be able to create one surface and always return that. ImageReader requires we make new Surfaces so our API *users* shouldn't cache the returned Surface.

File: shell/platform/android/io/flutter/embedding/engine/renderer/SurfaceTextureSurfaceProducer.java (hunk 6 above)
Comment (flar) [reply shell/platform/android/io/flutter/embedding/engine/renderer/SurfaceTextureSurfaceProducer.java:9]: Object.finalize() was deprecated in JDK9 and there are plans to remove it. See https://openjdk.org/jeps/421

Comment (flar) [reply shell/platform/android/io/flutter/embedding/engine/renderer/SurfaceTextureSurfaceProducer.java:9]: (The JEP points to alternatives with fewer issues)

Comment (johnmccutchan) [reply shell/platform/android/io/flutter/embedding/engine/renderer/SurfaceTextureSurfaceProducer.java:9]: Regardless, we rely on finalizers today so my comment still stands.

Comment (flar) [reply shell/platform/android/io/flutter/embedding/engine/renderer/SurfaceTextureSurfaceProducer.java:9]: I filed an issue to keep tabs on these uses: https://github.com/flutter/flutter/issues/141403

Comment (matanlurey) [reply shell/platform/android/io/flutter/embedding/engine/renderer/SurfaceTextureSurfaceProducer.java:12]: Done!

File: shell/platform/android/io/flutter/embedding/engine/renderer/SurfaceTextureSurfaceProducer.java (hunk 5 above)
Comment (matanlurey) [reply shell/platform/android/io/flutter/embedding/engine/renderer/SurfaceTextureSurfaceProducer.java]: Got it, thanks for clarifying. Done.

**Review by johnmccutchan — APPROVED** [review]

**Pull request merged**